features/steps/thens/alignment.py

```diff
@@ -172,3 +225,84 @@ def step_impl(context, inst):
                         yield ValidationOutcome(inst=inst, expected="same count of segments",
                                                 observed=observed_msg,
                                                 severity=OutcomeSeverity.ERROR)
+
+
+@gherkin_ifc.step(
+    'Each segment in the layout must have the same geometry type as its corresponding segment in the shape representation')
+def step_impl(context, inst):
+    # work back up the nesting tree to obtain the alignment
+    align = None
+    layout_entity_types = [
+        "IFCALIGNMENTHORIZONTAL",
+        "IFCALIGNMENTVERTICAL",
+        "IFCALIGNMENTCANT",
+    ]
+    for rel in inst.Nests:
+        layout = rel.RelatingObject
+        if layout.is_a().upper() in layout_entity_types:
+            for rel2 in layout.Nests:
+                align_ent = rel2.RelatingObject
+                if align_ent.is_a() == "IfcAlignment":
+                    align = ifc43x_alignment_validation.entities.Alignment().from_entity(align_ent)
+
+    if align is None:
+        msg = f"Error processing instance {str(inst)}. "
+        msg += "Expected an IfcAlignmentSegment nested 2 levels below an IfcAlignment."
+        raise ValueError(msg)
+
+    for idx, align_segment in enumerate(context.instances):
+
+        logic_segment = align_segment.DesignParameters
+        match logic_segment.is_a():
+            case "IfcAlignmentHorizontalSegment":
+                try:
+                    rep_segment = align.composite_curve.segments[idx].entity.ParentCurve
```

Comment:
The only difference between these cases seems to be align.composite_curve/gradient_curve/segmented_reference_curve. I would have probably written that as below, removing the need for a match stmt.

`getattr(align, {"IfcAlignmentHorizontalSegment": "composite_curve", ...}[logic_segment.is_a()]).segments[idx].entity.ParentCurve`

not so much because of the lines of code it saves per se, but because I get nervous about duplication and my inability to quickly see to what extent the branches are different.